### PR TITLE
Fixed Hard Coupling (com_users | plg_users_joomla)

### PR DESCRIPTION
**Pull request by carlitorweb**
Pull Request for Issue #20249

### Summary of Changes
Added `onContentPrepareForm` event to the plugin for doing the form require alterations to the passwords fields.

### Testing Instructions
Go to the administration and create a new user. Check the form that is shown need have the passwords fields with the attribute "require" if the plugin **User - Joomla!** have the Notification Mail to User option as No.

### Expected result
Passwords field with the "require" attribute 

### Actual result
Passwords field with the "require" attribute 

### Documentation Changes Required
No

### CODE REVIEW

File: plugins/user/joomla/joomla.php [hunk 1]
```diff
@@ -34,6 +34,40 @@ class PlgUserJoomla extends JPlugin
...
 	protected $db;
 
+	/**
+	 * Set as required the passwords fields when mail to user is set to No
+	 *
+	 * @param   JForm  $form  The form to be altered.
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
+	 *
```
Comment (zero-24) [inline plugins/user/joomla/joomla.php:50]: please add `@since  __DEPLOY_VERSION__` ;)

File: plugins/user/joomla/joomla.php [hunk 2]
```diff
@@ -34,6 +34,40 @@ class PlgUserJoomla extends JPlugin
...
+	 */
+	public function onContentPrepareForm($form, $data)
+	{
+		if (!($form instanceof JForm))
```
Comment (zero-24) [inline plugins/user/joomla/joomla.php]: hmm IMO this check can be dropped as if we get here without a valid form something else is broken anyway. I have not checked it but IIRC `$this->_subject->setError('JERROR_NOT_A_FORM');` would not work anyway.

File: plugins/user/joomla/joomla.php (hunk 1 above)
Comment (carlitorweb) [reply plugins/user/joomla/joomla.php:50]: always forget something :D

File: plugins/user/joomla/joomla.php [hunk 3]
```diff
@@ -34,6 +34,40 @@ class PlgUserJoomla extends JPlugin
 	 */
 	protected $db;
 
+	/**
+	 * Set as required the passwords fields when mail to user is set to No
+	 *
+	 * @param   JForm  $form  The form to be altered.
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
```
Comment (zero-24) [inline plugins/user/joomla/joomla.php:49]: It seams drone don't likes this line ;)

```
FILE: /drone/src/github.com/joomla/joomla-cms/plugins/user/joomla/joomla.php
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
--------------------------------------------------------------------------------
 43 | ERROR | Return comment requires a blank newline after it
```
http://ci.joomla.org/joomla/joomla-cms/5010/4

File: plugins/user/joomla/joomla.php (hunk 2 above)
Comment (carlitorweb) [reply plugins/user/joomla/joomla.php]: > if we get here without a valid form something else is broken anyway

I thought same, but I saw we use it inside another plugin, and no wanted to drop something without knowing the "exactly why"

**Comment by carlitorweb** [comment]
Done @zero-24 and thank you

File: plugins/user/joomla/joomla.php [hunk 4]
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
...
+	 *
+	 * @since  3.8
```
Comment (zero-24) [inline plugins/user/joomla/joomla.php]: Please use `__DEPLOY__VERSION__` this way the release script replace it with the correct version where that method get committed in :)

File: plugins/user/joomla/joomla.php [hunk 5]
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
...
 
+	/**
+	 * Set as required the passwords fields when mail to user is set to No
+	 *
+	 * @param   JForm  $form  The form to be altered.
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
+	 *
+	 * @since  __DEPLOY__VERSION__
```
Comment (Quy) [inline plugins/user/joomla/joomla.php]: Align. Add one more space.

**Comment by zero-24** [comment]
I have just tested this and found two issue. In case there is a validation error $data is no stdClass but just an empty array. And the second is what happen when the plugin is disabled. Currently the password field is not required. But it needs to be required than.

## How to reproduce (First Issue)

- set error reporting to max
- create a user with the username `test`
- try to create another user with the same username
- error pops up.
- when you debug you notice that the $data array is empty.

## How to reproduce (Second Issue)

- Disable the plugin
- try to create a user
- pw field is not required.

## How to solve the issues

- change default in com_users XML file that the pw's are always required
- restructure the plugin check so it disable the required status if running.

**Comment by carlitorweb** [comment]
>Disable the plugin

@zero-24 if you disabled the plugin, others things like the login logic will break, among other things. In fact, there is a string warning about this. In case this plugin is disabled because another plugin will handle User synchronization, then we no need more the "Notification Mail to User", and we no need check for this.

> try to create another user with the same username

We can add an `is_object()` check inside the condition for avoiding this

**Comment by carlitorweb** [comment]
I do not understand why the drone fail here?

**Comment by zero-24** [comment]
> In case this plugin is disabled because another plugin will handle User synchronization, then we no need more the "Notification Mail to User", and we no need check for this.

Well my point is that the password fields are than not required in that case ;)
So my suggestion was to change the logic so the fields are required allways but not required with the setting is enabeld ;)

But I'm also fine merge this without that change and when this got accepted i do a PR that would handle the proposed solution.

**Comment by carlitorweb** [comment]
@Quy can you give me a hand here? This code is marked as a conflict for the merge, was moved to the plugin for solved the issue. Or maybe I understand wrong this.

**Comment by Quy** [comment]
I think you have to update your branch since there were changes to `administrator/components/com_users/models/user.php` 4 days ago.

**Comment by carlitorweb** [comment]
Thank you @Quy

File: plugins/user/joomla/joomla.php [hunk 6]
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
...
 
+	/**
+	 * Set as required the passwords fields when mail to user is set to No
+	 *
+	 * @param   JForm  $form  The form to be altered.
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
+	 *
+	 * @since   __DEPLOY__VERSION__
```
Comment (SharkyKZ) [inline plugins/user/joomla/joomla.php]: It's `__DEPLOY_VERSION__` (one underscore between words).

File: plugins/user/joomla/joomla.php [hunk 7]
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
...
+	 */
+	public function onContentPrepareForm($form, $data)
+	{
+		// Check we are manipulating a valid user form before modifying it.
+		$name = $form->getName();
+
+		if ($name === 'com_users.user')
+		{
+			// Passwords fields are required when mail to user is set to No
+			if (is_object($data) && $data->id === 0 && $this->params->get('mail_to_user', '1') === '0')
```
Comment (SharkyKZ) [inline plugins/user/joomla/joomla.php]: Change `mail_to_user` param fallback value and comparison value to be integer and comparison to be not type-safe:
`$this->params->get('mail_to_user', 1) == 0` or `!$this->params->get('mail_to_user', 1)`.

**Comment by carlitorweb** [comment]
Thank you @SharkyKZ

**Comment by Quy** [comment]
Create a duplicate user.
An error message is displayed.
`Password` and `Confirm Password` are no longer marked required.

#14851 and #20313 could be related to this issue.

**Comment by SharkyKZ** [comment]
Data from user state is an array https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_users/models/user.php#L172.

This seems to be the case everywhere else.

**Comment by carlitorweb** [comment]
I made what @SharkyKZ suggest before:

> change default in com_users XML file that the pw's are always required
restructure the plugin check so it disable the required status if running.

Please, can you test again @Quy . Also, as the password field is required by default, we need to remove this, when we edit a user. 

PD: Sorry for the delay

**Comment by Quy** [comment]
On the backend, `Password` and `Confirm Password` are not required (right). After an error, these fields are marked required (wrong).

**Comment by carlitorweb** [comment]
>On the backend, `Password` and `Confirm Password` are not required (right)

So, change the pw´s fields required by default is wrong?

**Comment by Quy** [comment]
On the backend, leave the password blank for the system to auto generate the password. Without PR, the 2 password fields are not required. With PR, the password fields are not required initially, but after a duplicate error, they are marked required.

**Comment by carlitorweb** [comment]
@Quy now I got it. Sorry for be slow, sometime my English is not so good as I want it. 
So, with this PR and plugin User - Joomla, with the `Notification Mail to User` set on `No` passwords fields are required (Correct operation). But if you save a duplicate user, then passwords fields after get the error, are not required (Wrong operation, need stay as required).

Checking this, somehow $options['load_data'] come with the value false when you save a duplicate user, https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/libraries/src/MVC/Model/FormModel.php#L230 then, preprocessForm(...) trigger the plugin passing an empty $data, so that make the check inside onContentPrepareForm event fail.

**Comment by Quy** [comment]
Youre English fine to me. 👍 

Please read the following about password on the backend: https://github.com/joomla/joomla-cms/pull/20247#issuecomment-385167234

I was wrong with my report. I think your PR is fine, but I will have to test again to be sure.

**Comment by carlitorweb** [comment]
> Please read the following about password on the backend

Thank you

**Comment by SharkyKZ** [comment]
The issue is that on validation error returned data is an array. This could probably be fixed in core, but for now you can remove `is_object()` check and cast data as object.

**Comment by Quy** [comment]
Tested @ https://github.com/joomla/joomla-cms/pull/20275/commits/9e333a998656c0e97b979d43cf66d0d60cc501b1

> PHP Notice:  Undefined property: stdClass::$id in \plugins\user\joomla\joomla.php on line 58

**Comment by SharkyKZ** [comment]
@Quy how do you get this notice?

**Comment by Quy** [comment]
Check in the PHP error log file.

**Comment by SharkyKZ** [comment]
FormController never loads form data so whatever form data check is performed in plugin is not valid when saving.
[https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/libraries/src/MVC/Controller/FormController.php#L688](https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/libraries/src/MVC/Controller/FormController.php#L688)
Assuming this is intended and no changes should be made to models or controllers, see `System - Fields` plugin for a solution:
[https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/plugins/system/fields/fields.php#L274-L287](https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/plugins/system/fields/fields.php#L274-L287)

**Comment by carlitorweb** [comment]
Thank you @SharkyKZ and @Quy  .

**Comment by Quy** [comment]
Sorry I gave you wrong information prior. Here is the right information:

On the back end, when `Notification Mail to User` is `No`, then `Password` and `Confirm Password` should be marked with `*` and be required. 

Your PR, however, after an error, these 2 fields are no longer marked with `*`.

**Comment by SharkyKZ** [comment]
This isn't right yet. On form display, `$data` is always populated. It's normally an object but an array after returning from error. Data from input (`$jformData`) is not available on form display.

On save, `$data` is an empty array because `FormController` does not load data. So on save you need to use data from input (`$jformData`).

This is indeed #14851.

The same issue exists without PR. When `Notification Mail to User`  is set to `No` password fields appear to be required on form display but aren't validated on save. To test this, remove `required` attribute and class from markup using browser's developer tools and submit the form without passwords. The user will be saved.

**Comment by carlitorweb** [comment]
The real fix is try to handle this from the model or the controller. But I can bet will be B/C issue.

> Your PR, however, after an error, these 2 fields are no longer marked with *

Is true, I do not see this. However, if you try to save again, you get a warning telling you, the password field is required.

**Comment by carlitorweb** [comment]
@Quy I does not understand why, when we get the error the `*` gone. Even if you save, you got the warning saying the password is required, but no put the field as required in the HTML markup. Can you give me a hand with this? I try to find where I miss something, but no luck for now

**Comment by SharkyKZ** [comment]
It's because of the check in the plugin. Replace this:
```
if (is_array($data) && $jformData)
{
	$data = (object) $jformData;
}
```
With this:
```
if ($jformData && !$data) 
 { 
 	$data = $jformData; 
 } 
  
if (is_array($data)) 
{ 
 	$data = (object) $data; 
} 
```

**Comment by carlitorweb** [comment]
Thank you @SharkyKZ , that was But or I need close all here and go to watch a movie, or is almost the same I had in the check before :D . Anyway, thank you for the help

File: plugins/user/joomla/joomla.php [hunk 8]
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
...
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function onContentPrepareForm($form, $data)
+	{
+		// Check we are manipulating a valid user form before modifying it.
+		$name = $form->getName();
+
+		if ($name === 'com_users.user')
+		{
+			// In case there is a validation error (like duplicated user), $data is no a stdClass but just an empty array
```
Comment (Quy) [inline plugins/user/joomla/joomla.php]: Change `no` to `not`

File: plugins/user/joomla/joomla.php [hunk 9]
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
...
+				$data = $jformData;
+			}
+
+			if (is_array($data))
+			{
+				$data = (object) $data;
+			}
+
+			// Passwords fields are required when mail to user is set to No
+			if (isset($data->id) && (int) $data->id == 0 && !$this->params->get('mail_to_user', 1))
```
Comment (Quy) [inline plugins/user/joomla/joomla.php]: Can be simplified as follows: `if (empty($data->id) && !$this->params->get('mail_to_user', 1))`

**Comment by carlitorweb** [comment]
Done, thank @Quy

File: plugins/user/joomla/joomla.php [hunk 10]
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
...
+		{
+			// In case there is a validation error (like duplicated user), $data is not a stdClass but just an empty array
+			// TODO: Check inside the FormModel line 230, for put the right associated data for the form
+			$input = JFactory::getApplication()->input;
+			$jformData = $input->get('jform', array(), 'array');
+
+			if ($jformData && !$data)
+			{
+				$data = $jformData;
+			}
```
Comment (ggppdk) [inline plugins/user/joomla/joomla.php:65]: Can be written like:
```php
	if (!$data)
	{
		$data = JFactory::getApplication()->input->get('jform', array(), 'array');
	}
```

File: plugins/user/joomla/joomla.php [hunk 11]
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
...
+	 */
+	public function onContentPrepareForm($form, $data)
+	{
+		// Check we are manipulating a valid user form before modifying it.
+		$name = $form->getName();
+
+		if ($name === 'com_users.user')
+		{
+			// In case there is a validation error (like duplicated user), $data is not a stdClass but just an empty array
+			// TODO: Check inside the FormModel line 230, for put the right associated data for the form
```
Comment (mbabker) [inline plugins/user/joomla/joomla.php]: Really shouldn't make a reference to a specific line number in a code comment.  If the code's updated, this reference goes away.

Comment (carlitorweb) [reply plugins/user/joomla/joomla.php]: Is true, thank you @mbabker my mistake, sorry

File: plugins/user/joomla/joomla.php (hunk 10 above)
Comment (carlitorweb) [reply plugins/user/joomla/joomla.php:65]: look better, thank you

File: plugins/user/joomla/joomla.php (hunk 11 above)
Comment (SharkyKZ) [reply plugins/user/joomla/joomla.php]: The first comment should also be updated. `$data` is an empty array on save. After returning from error, `$data` is an array but populated.

@mbabker can you comment on form data issue? `FormController` does not load form data by default, making it unavailable in `onContentPrepareForm` when saving. Code based on `Form::getValue()` also fails. Is this correct behavior? Should we use data from input instead?

**Comment by Quy** [comment]
Related issue #17700 about empty $data.

**Comment by carlitorweb** [comment]
This is the correct way then to do this or need a different way?

**Comment by SharkyKZ** [comment]
The code works fine now.

Still thinking about @zero-24's suggestion to make password fields required by default. It would make sense not to auto-generate passwords when plugin is disabled, when mail to user option in plugin is disabled or when sending passwords is disabled in com_users.

**Comment by carlitorweb** [comment]
Ok, let do the suggestion of @zero-24

**Comment by brianteeman** [comment]
It never makes sense in the admin to make creating a password required.

**Comment by mbabker** [comment]
> It never makes sense in the admin to make creating a password required.

For creating a new account, I'd say it would make sense to default it to required and conditionally unrequire it based on the other settings.  Otherwise you're left in a spot where admin creates an account without setting a password and the new user's only way of getting into the account is through the "forgot my password" interface.  I don't think that's necessarily the best new user experience, but we all know what opinions are like.

**Comment by carlitorweb** [comment]
This is a valid argument imo [#20247 (comment)](https://github.com/joomla/joomla-cms/pull/20247#issuecomment-385167234)

**Comment by ghost** [comment]
@carlitorweb are the Test Instructions up-to-date?

**Comment by carlitorweb** [comment]
@Fran so far, yes

**Comment by ghost** [comment]
"Notification Mail to User"-option is "No", both Password-Fields have a "*". I'm confused about the Test Instructions. 

Also "Expected" and "Actual Result" have the same Description.

**Comment by SharkyKZ** [comment]
@franz-wohlkoenig Behavior before and after patch should be the same.

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on fe03395bce8b16d3db6685e938b7eb4fc7f228c6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20275">issues.joomla.org/tracker/joomla-cms/20275</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on b633bbfbdc99d9a8cf843ca6cae0e06c4c68d855<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20275">issues.joomla.org/tracker/joomla-cms/20275</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20275">issues.joomla.org/tracker/joomla-cms/20275</a>.</sub>

**Comment by HLeithner** [comment]
Thanks

**Pull request merged**